Fix part of #5734: Update Lint Reporter and update unknown ID flow

manas-yu opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes part of #5734

This PR introduces the following improvements:

- **Removed unnecessary fields**: Priority and summary have been removed from lint reports to streamline output.  
- **Standardized Issue IDs**: All Issue IDs are now logged in **UPPER_SNAKE_CASE** for consistency.  
- **Enhanced Unknown ID handling**:  
  - Issues with unknown IDs will still be included in the report without blocking the workflow.  
  - Developers are given the option to either update the exemption utilities for exemption support or resolve the issue directly.  
- **Stricter validation flow**:  
  - The script will **fail** if the report contains any of the following:  
    - Fatal issues  
    - Error issues  
    - Warning issues  
    - Unknown Issue IDs  
    - Redundant exemptions  
  - The script will **pass** only if the report contains **informational issues** exclusively.  

### Sample report output:
```shell
LINT CHECKS FAILED. Please fix the issues below.

Redundant Exemptions: 2
Fatal: 1
Error: 4
Warning: 1
Total Issues: 8

If you need additional help to resolve an issue, see https://googlesamples.github.io/android-custom-lint-rules/checks/severity.md.html

================================================================================
FILE: scripts/assets/android_lint_exemptions.textproto (2 issues)
================================================================================

Issue 1 of 2: REDUNDANT_EXEMPTION
  Severity: Redundant Exemption
  Line: 15
  Error Line: lint_issue_id: DUPLICATE_STRINGS
  Message: Redundant exemption found. Please remove it from the file.
  Explanation:
    In app/src/main/java/com/example/MainActivity.kt the 
    DUPLICATE_STRINGS exemption is redundant and can be 
    removed since there are no corresponding lint issues.
------------------------------------------------------------
Issue 2 of 2: REDUNDANT_EXEMPTION
  Severity: Redundant Exemption
  Line: 23
  Error Line: lint_issue_id: BACK_BUTTON
  Message: Redundant exemption found. Please remove it from the file.
  Explanation:
    In app/src/main/java/com/example/MainActivity.kt the 
    BACK_BUTTON exemption is redundant and can be removed 
    since there are no corresponding lint issues.

================================================================================
FILE: app/src/main/java/com/example/MainActivity.kt (3 issues)
================================================================================

Issue 1 of 3: CUSTOM_SPLASH_SCREEN (Category: Correctness)
  Severity: Fatal
  Line: 25
  Error Line: setTheme(R.style.SplashTheme)
  Message: Custom splash screen implementation may not work correctly on Android 12+
  Explanation:
    Starting with Android 12, the platform provides a new splash screen API that
    replaces custom splash screen implementations. Custom splash screens may not
    display correctly or may cause performance issues.

    Consider migrating to the new SplashScreen API provided by the
    androidx.core:core-splashscreen library for better compatibility and
    performance.
------------------------------------------------------------
Issue 2 of 3: AUTOFILL (Category: Security)
  Severity: Error
  Line: 45
  Error Line: <EditText android:inputType="textPassword" />
  Message: This field should specify an autofill hint
  Explanation:
    Password fields and other input fields should provide autofill hints to help
    autofill services provide better suggestions to users.

    Add android:autofillHints attribute to improve user experience.
------------------------------------------------------------
Issue 3 of 3: GRADLE_OVERRIDES (Category: Correctness)
  Severity: Warning
  Line: 12
  Message: Build configuration may be overridden by gradle.properties
  Explanation:
    Some build configurations in your build.gradle file may be overridden by
    properties defined in gradle.properties files. This can lead to unexpected
    build behavior.

================================================================================
FILE: app/src/main/res/layout/activity_main.xml (2 issues)
================================================================================

Issue 1 of 2: NEW_API (Category: Correctness)
  Severity: Error (FALSE POSITIVE)
  Line: 15
  Error Line: items.forEach { item ->
  Message: Call requires API level 24 (current min is 21): `java.lang.Iterable#forEach`
  Workaround:
    Use safeForEach from IterableExtensions.kt instead of directly
    calling forEach to avoid known lint false positives on API < 24.
------------------------------------------------------------
Issue 2 of 2: DUPLICATE_STRINGS (Category: Performance)
  Severity: Error
  Line: 8
  Message: String literal "Hello World" appears multiple times
  Explanation:
    The same string literal appears in multiple places. Consider extracting it
    to a string resource for better maintainability and localization support.

================================================================================
FILE: app/src/main/res/values/strings.xml (2 issues)
================================================================================

Issue 1 of 2: APP_BUNDLE_LOCALE_CHANGES (Category: Internationalization)
  Severity: Information
  Line: 23
  Message: This resource may be filtered out in App Bundle builds
  Explanation:
    When building App Bundles, locale-specific resources may be split into
    separate modules and not included in the base APK.

    Ensure that your app handles missing locale resources gracefully.
------------------------------------------------------------
Issue 2 of 2: MISSING_TRANSLATION (Category: Internationalization)
  Severity: Error
  Locations:
    1. File: app/src/main/res/values/strings.xml
       Line: 15
    2. File: app/src/main/res/values-es/strings.xml
       Line: 12
  Message: "app_name" is not translated in "fr" (French)
  Explanation:
    The default language is English. To support other languages, you should
    define the same string resources for those locales. Missing translations
    can lead to a poor user experience for users who have their device set to
    unsupported languages.

============================================================
UNUSED ENUM MAPPINGS DETECTED:
The following issue IDs are defined in issueIdMapping but no corresponding lint issues were found.
Please remove them from the LintIssueId enum and issueIdMapping:

  - KeyboardInaccessibleWidget -> KEYBOARD_INACCESSIBLE_WIDGET
  - LabelFor -> LABEL_FOR
  - ObsoleteSdkInt -> OBSOLETE_SDK_INT
  - RedundantLabel -> REDUNDANT_LABEL
  - Typos -> TYPOS

============================================================
ANDROID LINT CHECK FAILED
``` 

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

manas-yu committed:
update lint reporter

manas-yu committed:
ktlint

manas-yu committed:
update fail criterion

manas-yu committed:
issue id to upper snake case

manas-yu committed:
ktlint

manas-yu committed:
Merge branch 'develop' into update-lint-reporter

manas-yu committed:
fixx conflict

manas-yu commented:
Hi @bhavukJain PTAL, Thanks!

seanlip commented:
Hi @manas-yu, thanks -- I had a chat with Ben about this. Some notes:

- Change "Issues" to "issues" in the first line.
- Show the summary in the same order as the details. If you are showing the redundant exemptions info first then move it to the top of the summary.
- Will you still be colour-coding the output as you did in the demo?
- In the case of unknown issue IDs, treat them as regular issues (put them in the fatal/error/warning category, display the relevant info for them as the other errors/warnings/etc., and don't treat them out separately). Don't encourage or provide instructions for developers to augment the issueIdMapping. (This should be laid out on a wiki page if it isn't already, but you don't need to link to it here.)

Also, I clarified with Ben that, in general, the lint enum list should be a comprehensive list of the errors that we **actually encounter in the Android code**. So, if the lint check runs on the full codebase (not a subset), and an enum has no matching errors etc., then error at the end and ask the developer to drop the missing enums from the enum list. That way, we can push the list to be smaller over time.

I hope this makes sense, please ask if you have any questions about the above. Thanks!

manas-yu committed:
Merge branch 'develop' into update-lint-reporter

manas-yu committed:
remove unknown id reporting

manas-yu committed:
add tests

manas-yu committed:
wiki update

manas-yu committed:
Merge branch 'develop' into update-lint-reporter

manas-yu committed:
fix issue list

manas-yu committed:
ktlint

manas-yu committed:
remove mappings

manas-yu committed:
fix tests

manas-yu commented:
Hi @seanlip, PTAL at the updated sample output and let me know if any other changes are required. Thanks!

> Change "Issues" to "issues" in the first line.

Updated.

> Show the summary in the same order as the details. If you are showing the redundant exemptions info first then move it to the top of the summary.

Moved redundant exemptions to the top of the summary.

> Will you still be colour-coding the output as you did in the demo?

Yes, it's still there, and the output will be similarly color-coded in the terminal as showcased in the demo.

> In the case of unknown issue IDs, treat them as regular issues (put them in the fatal/error/warning category, display the relevant info for them as the other errors/warnings/etc., and don't treat them out separately). Don't encourage or provide instructions for developers to augment the issueIdMapping. (This should be laid out on a wiki page if it isn't already, but you don't need to link to it here.)

Treated the Unknown issue IDs as regular issues and included them in the report with no special info in the logging, and updated the Wiki instead with the instructions.

> So, if the lint check runs on the full codebase (not a subset), and an enum has no matching errors etc., then error at the end and ask the developer to drop the missing enums from the enum list. That way, we can push the list to be smaller over time.

Made the script to error at the end if there are issue IDs present in the mapping that don't come up in the full project report, prompting the user to remove them from the enum list and the issue ID mapping.

seanlip commented:
Thanks Manas. Looks good! Just one suggestion: for the redundant exemptions bit, I suggest you format them similarly to the others. For example (wrap the following as needed):

```
Redundant Exemptions: 2
Fatal: 1
Error: 4
Warning: 1
Total Issues: 8

If you need additional help to resolve an issue, see https://googlesamples.github.io/android-custom-lint-rules/checks/severity.md.html


================================================================================
FILE: scripts/assets/android_lint_exemptions.textproto (2 issues)
================================================================================

Issue 1 of 2: Redundant exemption found
  Severity: Redundant exemption
  Line: ...
  Error Line: ...
  Message: Redundant exemption found. Please remove it from the file.
  Explanation:
    In app/src/main/java/com/example/MainActivity.kt the DUPLICATE_STRINGS exemption is redundant and can be removed since there are no corresponding lint issues.

Issue 2 of 2: Redundant exemption found
  Severity: Redundant exemption
  Line: ...
  Error Line: ...
  Message: Redundant exemption found. Please remove it from the file.
  Explanation:
    In app/src/main/java/com/example/MainActivity.kt the BACK_BUTTON exemption is redundant and can be removed since there are no corresponding lint issues.

```

Otherwise, it looks great to me. Thanks!

manas-yu committed:
update logging

manas-yu committed:
update test

manas-yu committed:
update wiki

manas-yu commented:
Thanks @seanlip updated redundant exemption logging, PTAL.

seanlip commented:
Thanks Manas. In the redundant exemption messages:

- Can you include the line number and error line (similar to the other messages)?
- Make sure the explanation section is wrapped similarly to the other errors. The current line length seems too long.


Thanks!

manas-yu committed:
wrap explanation

manas-yu committed:
add test

manas-yu committed:
update test

manas-yu committed:
ktlint

manas-yu commented:
Hi @seanlip PTAL, addressed comments.

> Can you include the line number and error line (similar to the other messages)?

Included line number and Error line logging.

 > Make sure the explanation section is wrapped similarly to the other errors. The current line length seems too long.
 
 Fixed. Thanks!

seanlip commented:
@manas-yu Looks great to me. Thanks!

Just one note, let's change the severity of the first two issues to "Redundant Exemption" instead of "Warning", so they match the index at the top. Otherwise, no concerns!

manas-yu committed:
Merge branch 'develop' into update-lint-reporter

manas-yu committed:
change severity

manas-yu commented:
> change the severity of the first two issues to "Redundant Exemption" instead of "Warning"

Updated, assigning for final confirmation. Thanks!

manas-yu committed:
remove id entries

seanlip commented:
Thanks @manas-yu. LGTM for product!

seanlip commented:
P.S. @manas-yu assigning you due to merge conflicts.

manas-yu committed:
Merge branch 'develop' into update-lint-reporter

manas-yu committed:
Update android_lint_exemptions.textproto

manas-yu commented on scripts/assets/android_lint_exemptions.textproto:
```diff
@@ -1024,6 +1018,22 @@ android_lint_exemption: {
   exempted_file_path: "app/src/main/res/layout-land/onboarding_slide.xml"
   lint_issue_id: UNUSED_RESOURCES
 }
+android_lint_exemption: {
```
These are just for conflict resolution and will be removed in the chained PR #5936.

manas-yu commented:
Hi @adhiamboperes assigning you as well since this is a big PR and needs to get merged for evaluation, Bhavuk is already taking a pass.

Cc: @bhavukJain

bhavukJain commented on scripts/src/java/org/oppia/android/scripts/lint/AndroidLintRunner.kt:
```diff
@@ -111,7 +112,8 @@ class AndroidLintAnalyzer(
       exemptionProtoPath = exemptionProtoPath,
-      groupByIssueSeverity = groupByIssueSeverity
+      groupByIssueSeverity = groupByIssueSeverity,
+      reportUnusedEnum
```
Should we use named parameters for reportUnusedEnum as well?

manas-yu replied:
Done.

bhavukJain commented on scripts/src/java/org/oppia/android/scripts/lint/LintAnalysisReporter.kt:
```diff
@@ -128,8 +121,15 @@ private data class CacheEntry(
+
 /** Reporter class for analyzing XML lint reports and extracting issues. */
-class LintAnalysisReporter {
+class LintAnalysisReporter(private val repoRoot: File) {
```
Nit: Update class level documentation to reflect that this is now accepting a parameter as well.

manas-yu replied:
Done

bhavukJain commented on scripts/src/java/org/oppia/android/scripts/lint/LintAnalysisReporter.kt:
```diff
@@ -394,23 +369,123 @@ class LintAnalysisReporter {
-  fun printLintReport(issues: List<LintIssue>, groupByIssueSeverity: Boolean) {
-    printSeveritySummary(issues)
+  fun printLintReport(
+    issues: List<LintIssue>,
```
This is a bit confusing, i.e. having both issues and allIssues. Can we rename it to filteredIssues and allIssues to make the distinction clearer?

manas-yu replied:
Thanks renamed.

bhavukJain commented on scripts/src/java/org/oppia/android/scripts/lint/LintAnalysisReporter.kt:
```diff
@@ -394,23 +369,123 @@ class LintAnalysisReporter {
+   * @param indent Indentation to apply to each line
+   * @return Formatted text with proper indentation and line wrapping
+   */
+  private fun wrapText(text: String, width: Int = 70, indent: String = "    "): String {
```
This method seems to be doing too much, parsing paragraphs, word wrapping, and indentation. Can we break this down into smaller focussed methods?

manas-yu replied:
Thanks! Split the function into: 
- `wrapParagraph`: handles word wrapping for one paragraph
- `wrapText`: orchestrates paragraphs + indentation

bhavukJain commented on scripts/src/java/org/oppia/android/scripts/proto/android_lint.proto:
```diff
@@ -33,42 +33,27 @@ enum LintIssueId {
   CUSTOM_SPLASH_SCREEN = 5;
   DUPLICATE_STRINGS = 6;
   GRADLE_OVERRIDES = 7;
-  IMPLIED_QUANTITY = 8;
```
Should we deprecate them first before removing to maintain backward compatibility? Hope removing them directly would be safe.

manas-yu replied:
Since the removed issue IDs are already absent from the codebase and no existing exemptions reference them, it should be safe to remove them directly without deprecation. We could alternatively mark them as deprecated (e.g., `NOTIFY_DATA_SET_CHANGED = 18 [deprecated = true];`), but the intent here was to keep the list limited to only the issues currently reported in the codebase, hence I removed them. Let me know if deprecation would be preferred instead.

adhiamboperes replied:
I think it is fine to not add this change.

manas-yu committed:
Merge branch 'develop' into update-lint-reporter

manas-yu committed:
address comments

manas-yu commented:
Hi @adhiamboperes, PTAL addressed comments.

Cc: @bhavukJain

manas-yu committed:
fix kdoc

adhiamboperes reviewed (approved):
LGTM, thanks! Resolved @bhavukJain's comments to unblock merge.
